# Resources

* EBLOSUM{40,62,80} - BLOSUM matrix obtained from ClostalW [Clustal]
* EPAM{120,250,350} - PAM matrix obtained from ClostalW [Clustal]
* {BBS_,RV}*.btq - binary transition-quantity generated from from BAliBASE database [BAliBASE]
* homstrad*.btq - binary transition-quantity generated from homstrad database [Homstrad]
* MRAMA* - Ramachandran matrix [RAMA]
* prefab4*.btq - binary transition-quantity generated from PREFAB database [Prefab]
* SABmark*.btq - binary transition-quantity generated from SABmark database [SABmark]
* VTML200I - VTML matrix [VTML]

# References
* [Clustal] http://www.clustal.org/clustal2/
* [BAliBASE] http://www.lbgi.fr/balibase/
* [Homstrad] http://mizuguchilab.org/homstrad/
* [RAMA] 東京理科大学理工学専攻情報科学科修士論文, 6306631, 中村卓, タンパク質の立体構造と各アミノ酸のペアの起こりやすさを考慮したアライメントアルゴリズム
* [Prefab] https://www.drive5.com/muscle/prefab.htm
* [SABmark] https://academic.oup.com/bioinformatics/article/21/7/1267/268759/SABmark-a-benchmark-for-sequence-alignment-that
* [VTML] Mueller T, Spang R, Vingron M. Estimating amino acid substitution models: a comparison of Dayhoff’s estimator, the resolvent approach and a maximum likelihood method. Mol. Biol. Evol. 2002;19:8–13.